pandas/core/frame.py

```diff
@@ -808,6 +808,21 @@ def __init__(
             if len(data) > 0:
                 if is_dataclass(data[0]):
                     data = dataclasses_to_dicts(data)
+
+                # Check if all elements in data are Series with categorical indices
+                if all(
```

Comment:
This fix is too specific. This should be fixed where the Index[object] is being created

Reply:
Indeed, I already see tests failing due to ensure the categorical indices like this.

Reply:
Hi @mroeschke, I've been digging and I've found out that the following example actually works:

```
s1 = pd.Series([1, 2], index=pd.CategoricalIndex(["a", "b"]))
s2 = pd.Series([3, 4], index=pd.CategoricalIndex(["b", "a"]))
pd.DataFrame([s1, s2]).columns
```
Returning:

`CategoricalIndex(['a', 'b'], categories=['a', 'b'], ordered=False, dtype='category')
`

It turns out that `union` (in `Index` class), when comparing both objects `dtype` also includes the `categories`, and if we use the issue's example, two `CategoricalIndex` objects with different `categories` turn `self.dtype != other.dtype` to be True. And following the code, this leads to a change of their `dtypes` transforming them to `Index`, and recursively calling `union` again, to unite them as an `Index` object. If both objects have the same categories, as the example I shared above, the return object is a `CategoricalIndex`. 

Since it seems that `CategoricalIndex` objects are not properly handled in `union`. My proposal here would be to add a condition where `self` and `other` are properly dealt when both are `CategoricalIndex`. Wdyt? Additional unknown changes may need to be done following this path.